src/lib/comms/sol-socket-impl-linux.c

```diff
@@ -437,85 +361,98 @@ sol_socket_linux_bind(struct sol_socket *socket, const struct sol_network_link_a
-{
-    switch (option) {
-    case SOL_SOCKET_OPTION_REUSEADDR:
-        return SO_REUSEADDR;
-    case SOL_SOCKET_OPTION_REUSEPORT:
-        return SO_REUSEPORT;
-    default:
-        SOL_WRN("Invalid option %d", option);
-        break;
-    }
-
-    return -1;
-}
-
-static int
-sol_socket_level_to_linux(enum sol_socket_level level)
+sol_socket_linux_set_read_monitor(struct sol_socket *s, bool on)
 {
-    switch (level) {
-    case SOL_SOCKET_LEVEL_SOCKET:
-        return SOL_SOCKET;
-    case SOL_SOCKET_LEVEL_IP:
-        return IPPROTO_IP;
-    case SOL_SOCKET_LEVEL_IPV6:
-        return IPPROTO_IPV6;
-    default:
-        SOL_WRN("Invalid level %d", level);
-        break;
-    }
+    bool ret;
+    struct sol_socket_linux *sock = (struct sol_socket_linux *)s;
 
-    return -1;
-}
+    SOL_NULL_CHECK(sock->on_can_read, -EINVAL);
 
-static int
-sol_socket_linux_setsockopt(struct sol_socket *socket, enum sol_socket_level level,
-    enum sol_socket_option optname, const void *optval, size_t optlen)
-{
-    int l, option;
-    struct sol_socket_linux *s = (struct sol_socket_linux *)socket;
+    if (on == sock->read_monitor)
+        return 0;
 
-    l = sol_socket_level_to_linux(level);
-    option = sol_socket_option_to_linux(optname);
+    sock->read_monitor = on;
+    if (on)
+        ret = sol_fd_set_flags(sock->watch,
+            sock->write_monitor ? SOL_FD_FLAGS_OUT | SOL_FD_FLAGS_IN : SOL_FD_FLAGS_IN);
+    else
+        ret = sol_fd_unset_flags(sock->watch, SOL_FD_FLAGS_IN);
 
-    return setsockopt(s->fd, l, option, optval, optlen);
+    return ret ? 0 : -EBADF;
 }
 
 static int
-sol_socket_linux_getsockopt(struct sol_socket *socket, enum sol_socket_level level,
-    enum sol_socket_option optname, void *optval, size_t *optlen)
+sol_socket_linux_set_write_monitor(struct sol_socket *s, bool on)
 {
-    int ret, l, option;
-    socklen_t len = 0;
-    struct sol_socket_linux *s = (struct sol_socket_linux *)socket;
+    bool ret;
+    struct sol_socket_linux *sock = (struct sol_socket_linux *)s;
 
-    option = sol_socket_option_to_linux(optname);
-    l = sol_socket_level_to_linux(level);
+    SOL_NULL_CHECK(sock->on_can_write, -EINVAL);
 
-    ret = getsockopt(s->fd, l, option, optval, &len);
-    SOL_INT_CHECK(ret, < 0, ret);
+    if (on == sock->write_monitor)
+        return 0;
 
-    *optlen = len;
-    return 0;
+    sock->write_monitor = on;
+    if (on)
+        ret = sol_fd_set_flags(sock->watch,
+            sock->read_monitor ? SOL_FD_FLAGS_OUT | SOL_FD_FLAGS_IN : SOL_FD_FLAGS_OUT);
+    else
+        ret = sol_fd_unset_flags(sock->watch, SOL_FD_FLAGS_OUT);
+
+    return ret ? 0 : -EBADF;
 }
 
-const struct sol_socket_impl *
-sol_socket_get_impl(void)
+struct sol_socket *
+sol_socket_default_new(const struct sol_socket_default_options *options)
 {
-    static const struct sol_socket_impl impl = {
+    int ret;
+    struct sol_socket_linux *s;
+    int fd, socktype = SOCK_CLOEXEC | SOCK_NONBLOCK | SOCK_DGRAM;
```

Comment:
`SOCK_DGRAM` shouldn't be here. It will be a pain to remember to change this when other types are enabled.